File: 7 - Processamento de Imagens/Atividades Práticas/Atividade03/atividade.py
import numpy as np
import cv2
from matplotlib import pyplot as plt

# Carregar a imagem do disco
img_gray = cv2.imread('imagem.jpg')

# Calcula o histograma não normalizado da imagem
counts, bins = np.histogram(img_gray.flatten(), 256, [0, 256])

# Calcula a função de distribuição cumulativa (CDF)
cdf = counts.cumsum()
cdf_normalized = cdf * counts.max() / cdf.max()

# Realiza a transformação dos níveis de intensidade (equalização)
cdf_m = np.ma.masked_equal(cdf, 0)
cdf_m = (cdf_m - cdf_m.min()) * 255 / (cdf_m.max() - cdf_m.min())
cdf = np.ma.filled(cdf_m, 0).astype('uint8')
img_eq = cdf[img_gray]

# Calcula o histograma da imagem equalizada
counts_eq, bins_eq = np.histogram(img_eq.flatten(), 256, [0, 256])

# Exibindo a imagem original e a imagem equalizada lado a lado
plt.figure(figsize=(18, 12))

plt.subplot(221)
plt.imshow(img_gray, cmap='gray')
plt.title('Original')

plt.subplot(222)
plt.imshow(img_eq, cmap='gray')
plt.title('Equalizada')

# Plotar o gráfico de T(r) na terceira parte
plt.subplot(223)
plt.plot(cdf_normalized, color='b')
plt.hist(img_gray.flatten(), 256, [0,256], color='r', alpha=0.5)
plt.xlim([0, 255])
plt.legend(('CDF','Histograma Original'), loc = 'upper left')
plt.title('Função de Mapeamento T(r) e Histograma Original')

# Plotando o histograma da imagem equalizada
plt.subplot(224)
plt.hist(img_eq.flatten(), 256, [0, 256], color='g', alpha=0.5)
plt.xlim([0, 255])
plt.title('Histograma Equalizado')

# Salva a imagem equalizada no disco
cv2.imwrite('imagem_equalizada.jpg', img_eq)

plt.show()

File: 7 - Processamento de Imagens/Atividades Práticas/Atividade03/atividade.m
img_gray = imread('imagem.jpg'); % Ler a imagem do disco

% Calcula o histograma não normalizado da imagem
[counts, ~] = imhist(img_gray);

% Calcula a função de distribuição cumulativa (CDF)
cdf = cumsum(counts) / numel(img_gray);

% Realiza a transformação dos níveis de intensidade
img_eq = cdf(double(img_gray)+1) * 255;

% Converte a imagem equalizada para o tipo de dados original
img_eq = uint8(img_eq);

% Exibindo a imagem original e a imagem equalizada lado a lado
figure; % Cria uma nova figura

subplot(2,2,1); % Ajusta para a primeira imagem na posição superior esquerda
imshow(img_gray);
title('Original');

subplot(2,2,2); % Ajusta para a segunda imagem na posição superior direita
imshow(img_eq);
title('Equalizada');

% Plotar o gráfico de T(r) na parte inferior esquerda
subplot(2,2,3);
plot((0:255), cdf*255);
axis([0 255 0 255]);
xlabel('Valores de intensidade originais');
ylabel('Valores de intensidade equalizados');
title('Função de Mapeamento T(r)');

% Calcula e plota o histograma equalizado na parte inferior direita
[counts_eq, ~] = imhist(img_eq); % Calcula o histograma da imagem equalizada
subplot(2,2,4); % Ajusta para plotar o histograma equalizado
bar((0:255), counts_eq, 'k');
axis([0 255 0 max(counts_eq)]); % Ajusta os eixos para melhor visualização
title('Histograma Equalizado');
xlabel('Valores de intensidade');
ylabel('Quantidade de Pixels');

% Salva a imagem equalizada no disco
imwrite(img_eq, 'imagem_equalizada.jpg');

